File: 2015/go/05.go
// Advent of Code 2015, Day 5
// (c) aichingert

package main

import (
	"fmt"
	"os"
	"strings"
)

// You can't create a const array in go :I
var I = []string{"ab", "cd", "pq", "xy"}

func part1(lines []string) int {
	c := 0

	for _,s := range lines {
		vc := 0
		if is_vowel(s[len(s)-1]) { vc += 1 }
		v := true
		dd := false

		for i:=0 ; i < len(s)-1 ; i++ {
			if is_vowel(s[i]) { vc += 1 }
			if s[i] == s[i+1] { dd = true }
			if contains(string(s[i])+string(s[i+1])) {
				v = false
				break
			}
		}

		if v && dd && vc > 2 { c += 1 }
	}

	return c
}

func is_vowel(ch byte) bool {
	return ch == 97 || ch == 101 || ch == 105 || ch == 111 || ch == 117
}

func contains(ch string) bool {
	for i:=0 ; i < len(I) ; i++ {
		if ch == I[i] { return true }
	}
	return false
}

func part2(lines []string) int {
	c := 0

	for _,s := range lines {
		f := false
		t := false

		for i := 0 ; i < len(s)-2 ; i++ {
			for j := i+2 ; j < len(s)-1 ; j++ {
				if s[i] == s[j] && s[i+1] == s[j+1] { f = true }
			}

			if s[i] == s[i+2] { t = true }
		}

		if f && t { c += 1 }
	}

	return c
}

func main() {
	dat, err := os.ReadFile("../input/05")

	if err != nil {
		panic(err)
	}

	lines := strings.Split(string(dat), "\n")
	fmt.Printf("Part 1: %d\n", part1(lines))
	fmt.Printf("Part 2: %d\n", part2(lines))
}

File: 2015/python/05.py
# Advent of Code 2015, day 5
# (c) aichingert

I = ["ab","cd","pq","xy"]

def part1(lines):
    c = 0
    for s in lines:
        if s[-1] == 'a' or s[-1] == 'e' or s[-1] == 'i' or s[-1] == 'o' or s[-1] == 'u': vc = 1
        else: vc = 0
        v = True
        dd = False 

        for i in range(len(s)-1):
            if s[i] == 'a' or s[i] == 'e' or s[i] == 'i' or s[i] == 'o' or s[i] == 'u':
                vc += 1
            if s[i:i+2] in I:
                v = False
                break
            if s[i] == s[i+1]:
                dd = True

        if v and dd and vc >= 3:
            c += 1
    return c

def part2(lines):
    c = 0
    for s in lines:
        f = False
        t = False

        for i in range(len(s)-2):
            for j in range(i+2,len(s)-1):
                if (s[i] == s[j] and s[i+1] == s[j+1]):
                    f = True
            if s[i] == s[i+2]:
                t = True

        if f and t:
            c += 1
    return c

inp = open("../input/05").read().split('\n')

print("Part 1:",part1(inp))
print("Part 2:",part2(inp))
